tensorflow/compiler/xla/service/llvm_ir/llvm_util.cc

```diff
@@ -93,6 +93,59 @@ llvm::CallInst* EmitCallToIntrinsic(
   return b->CreateCall(intrinsic, AsArrayRef(operands));
 }
 
+// Return NVPTX,AMDGPU intrinsic
+GPUIntrinsics GetIntrinsic(TargetIntrinsicID intrinsic_id) {
+
+  CHECK_LE(intrinsic_id, kLast_id);
+  const GPUIntrinsics intrinsic_array[] =
+      {
+        [kShfl_down_f32] = {llvm::Intrinsic::nvvm_shfl_sync_down_f32,
+                            llvm::Intrinsic::not_intrinsic},
+        [kShfl_down_i32] = {llvm::Intrinsic::nvvm_shfl_sync_down_i32,
+                            llvm::Intrinsic::not_intrinsic},
+        [kThread_id_x] = {llvm::Intrinsic::nvvm_read_ptx_sreg_tid_x,
+                          llvm::Intrinsic::amdgcn_workitem_id_x},
+        [kThread_id_y] = {llvm::Intrinsic::nvvm_read_ptx_sreg_tid_y,
+                          llvm::Intrinsic::amdgcn_workitem_id_y},
+        [kThread_id_z] = {llvm::Intrinsic::nvvm_read_ptx_sreg_tid_z,
+                          llvm::Intrinsic::amdgcn_workitem_id_z},
+        [kBlock_id_x] = {llvm::Intrinsic::nvvm_read_ptx_sreg_ctaid_x,
+                         llvm::Intrinsic::amdgcn_workgroup_id_x},
+        [kBlock_id_y] = {llvm::Intrinsic::nvvm_read_ptx_sreg_ctaid_y,
+                         llvm::Intrinsic::amdgcn_workgroup_id_y},
+        [kBlock_id_z] = {llvm::Intrinsic::nvvm_read_ptx_sreg_ctaid_z,
+                         llvm::Intrinsic::amdgcn_workgroup_id_z},
+        [kBarrier_id] = {llvm::Intrinsic::nvvm_barrier0,
+                         llvm::Intrinsic::amdgcn_s_barrier},
+        [kLast_id]    = {llvm::Intrinsic::not_intrinsic,
+                         llvm::Intrinsic::not_intrinsic}
+      };
+  return intrinsic_array[intrinsic_id];
```

Comment:
Could we make this a switch statement?  That would be more idiomatic in our code.  It would also avoid the footgun of forgetting an element in this list (because the switch will warn if we don't enumerate all possibilities).  It would also let us CHECK-fail on kLastId being passed to this function, which seems good.